### PR TITLE
Wishing Compass Solver, Metal Detector Solver improvements, add /neudiag, other misc changes

### PR DESCRIPTION
**Pull request by CraftyOldMiner**
- Added Wishing Compass Solver that shows target coordinates using two samples. The text can be clicked to add a Skytils waypoint if Skytils is present.

- Lots of metal detector logic improvements, including solving using known locations based on Keeper coordinates. This usually only requires a single position. This also addresses cases where a solution cannot be found when Mines of Divan generation is glitched or people intentionally add/remove blocks around chests.

- Crystal Hollows crystal states are now updated when the Heart of the Mountain menu is opened.

- Added /neudiag command and `NEUDebugLogger` to support dynamic enabling/disabling of debug logging and dumping of diagnostic data for support/debugging purposes. Metal detector and wishing compass currently are supported, with the metal detector data being sufficient to investigate sporadic failures to find a solution.

- Added a Line utility class to support the Wishing Compass Solver.

- Various small `MiningOverlay` fixes, tweaks, and clarified naming.

- Minor name tweaks for increased clarity.

- Renamed CraftyOldMiner to Keebler408

### CODE REVIEW

File: src/main/java/io/github/moulberry/notenoughupdates/NotEnoughUpdates.java [hunk 1]
```diff
@@ -24,20 +24,8 @@
...
-import io.github.moulberry.notenoughupdates.miscfeatures.FishingHelper;
-import io.github.moulberry.notenoughupdates.miscfeatures.ItemCooldowns;
-import io.github.moulberry.notenoughupdates.miscfeatures.ItemCustomizeManager;
-import io.github.moulberry.notenoughupdates.miscfeatures.MiningStuff;
-import io.github.moulberry.notenoughupdates.miscfeatures.NullzeeSphere;
-import io.github.moulberry.notenoughupdates.miscfeatures.PetInfoOverlay;
-import io.github.moulberry.notenoughupdates.miscfeatures.SlotLocking;
-import io.github.moulberry.notenoughupdates.miscfeatures.StorageManager;
-import io.github.moulberry.notenoughupdates.miscfeatures.SunTzu;
+import io.github.moulberry.notenoughupdates.miscfeatures.*;
```
Comment (IRONM00N) [inline src/main/java/io/github/moulberry/notenoughupdates/NotEnoughUpdates.java:27]: me no like

File: src/main/java/io/github/moulberry/notenoughupdates/options/seperateSections/Mining.java [hunk 2]
```diff
@@ -1,710 +1,718 @@
...
+	@Expose
+	@ConfigOption(
+		name = "Replace Mithril sounds",
+		desc = "Replace the break sounds of mithril and titanium in the Dwarven mines. Requires a texture pack with this feature"
+	)
+	@ConfigAccordionId(id = 7)
+	@ConfigEditorBoolean
+	public boolean mithrilSounds = false;
+
+}
```
Comment (IRONM00N) [inline src/main/java/io/github/moulberry/notenoughupdates/options/seperateSections/Mining.java:718]: why is every line changed?

File: src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java [hunk 3]
```diff
@@ -298,48 +237,43 @@ public void update() {
...
+				if (forges && cleanName.startsWith(" ") && profileConfig != null) {
+					char firstChar = cleanName.trim().charAt(0);
 					if (firstChar < '0' || firstChar > '9') {
 						forges = false;
 					} else {
 
 						if (name.contains("LOCKED")) {
 							ForgeItem item = new ForgeItem(forgeInt, 1, true);
-							replaceForgeOrAdd(item, hidden.forgeItems, true);
+                            replaceForgeOrAdd(item, profileConfig.forgeItems, true);
```
Comment (IRONM00N) [inline src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java:272]: fix formatting, please

File: src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java [hunk 4]
```diff
@@ -298,48 +237,43 @@ public void update() {
...
 						} else if (name.contains("EMPTY")) {
 							ForgeItem item = new ForgeItem(forgeInt, 0, true);
-							replaceForgeOrAdd(item, hidden.forgeItems, true);
-							//forgeStringsEmpty.add(DARK_AQUA+"Forge "+ Utils.trimIgnoreColour(name).replaceAll("\u00a7[f|F|r]", ""));
+                            replaceForgeOrAdd(item, profileConfig.forgeItems, true);
```
Comment (IRONM00N) [inline src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java:275]: ditto

File: src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java [hunk 5]
```diff
@@ -467,8 +386,8 @@ public void update() {
 						overlayStrings.add(gemstonePowder);
 						break;
 					case 3:
-						if (hidden != null) {
-							overlayStrings.addAll(getForgeStrings(hidden.forgeItems));
+						if(profileConfig != null) {
```
Comment (IRONM00N) [inline src/main/java/io/github/moulberry/notenoughupdates/overlays/MiningOverlay.java:389]:
```suggestion
						if (profileConfig != null) {
```

**Review by IRONM00N — CHANGES_REQUESTED** [review]
see comments I made

**Comment by Keebler408** [comment]
@IRONM00N 
> see comments I made
Addressed. What are your thoughts on adding a .gitattributes and .idea/codeStyles/Project.xml to the project to prevent some of these issues in the future?

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/CrystalWishingCompassSolver.java [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+		solution = null;
+	}
+
+	@SubscribeEvent
+	public void onWorldLoad(WorldEvent.Unload event) {
+		resetForNewTarget();
+		isSkytilsPresent = Loader.instance().getActiveModList().stream().anyMatch(mod -> mod
+			.getName()
+			.toLowerCase()
+			.contains("skytils"));
```
Comment (NopoTheGamer) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/CrystalWishingCompassSolver.java:71]: cant you do `Loader.isModLoaded("skytils")` ?

**Review by NopoTheGamer — APPROVED** [review]

**Review by Lulonaut — APPROVED** [review]

**Comment by IRONM00N** [comment]
> @IRONM00N 
> 
> Comments have been addressed.
> 
> 
> 
> What are your thoughts on adding a .gitattributes to the project to prevent the line ending issues that were seen on Mining.java in the future? Adding .gitattributes would, of course, be in a separate PR. See https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/ for details if you need them.

I would appreciate it, I didn't realize that we never added a .gitattributes yet.

**Review by IRONM00N — APPROVED** [review]

**Pull request merged**